src/lib/inputs/InputNumber.svelte

```diff
@@ -0,0 +1,397 @@
+<script lang="ts">
+    import { createEventDispatcher, onMount } from "svelte";
+
+    // Props
+    export let variant: "default" | "line" | "outline" = "default";
+    export let size: "xsmall" | "small" | "medium" | "large" | "xlarge" =
+        "medium";
+    export let background: string = "white";
+    export let color: string = "#000";
+    export let value: number = 0;
+    export let style: string = "";
+    export let label: boolean = false;
+    export let labelColor: string = "#000";
+    export let disabled: boolean = false;
+    export let isError: boolean = false;
+    export let isLoading: boolean = false;
+    export let min: number = -10000;
+    export let max: number = 10000;
+    export let step: number | "any" = 1;
+
+    // Event dispatcher
+    const dispatch = createEventDispatcher();
+
+    // Variables
+    let containerClassList = ["input"];
+    let containerClassString = "";
+    let readOnly: boolean = false;
+
+    // Lifecycle Hooks
+    onMount(() => {
+        containerClassList.push(size);
+        containerClassList.push(variant);
+        if (isLoading) {
+            containerClassList.push("loading");
+        } else if (isError) {
+            containerClassList.push("error");
+        }
+        containerClassString = containerClassList.join(" ");
+
+        if (value < min) {
+            value = min;
+        } else if (value > max) {
+            value = max;
+        }
+    });
+
+    // Reactive Statements
+    $: {
+        containerClassList = ["input", size, variant];
+        if (isLoading) {
+            containerClassList.push("loading");
+        } else if (isError) {
+            containerClassList.push("error");
+        }
+        containerClassString = containerClassList.join(" ");
+
+        readOnly = disabled || isLoading || isError;
+    }
+
+    $: inputStyle = `
+  background: ${disabled || isError || isLoading ? "#ccc" : background};
+  --color: ${color};
+  color: ${color};
+  ${style}
+`;
+
+    // Handle input event
+    const handleInput = (e: Event) => {
+        if (e.target instanceof HTMLInputElement) {
+            let inputValue = parseInt(e.target.value);
+
+            if (inputValue < min) {
+                inputValue = min;
+            } else if (inputValue > max) {
+                inputValue = max;
+            }
+
+            value = inputValue;
+            e.target.value = value.toString();
+            dispatch("input", e);
+        }
+    };
+
+    // Increment Value
+    const incrementValue = () => {
+        if (isLoading || isError || disabled) return;
```

Comment:
do you need to add readonly to all the conditions for these functions to return and not execute?

Reply:
Yes, otherwise a user could fruitlessly alter input while there's already an error happening. Feels like more of a UX feature to me.

Reply:
Unless I just said if (isLoading || isError) disabled = true
But either way I don't think that a user should be able to alter the text while in a loading or error state.